module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php

```diff
@@ -282,6 +282,40 @@ public function getCleanISBN()
         return $isbn13;
     }
 
+    /**
+     * Return all valid ISBNs found in the record.
+     *
+     * @param array $flags Flags to prefer ISBN-10 or ISBN-13. If 'only10' is true,
+     * only ISBN-10s will be returned. If 'normalize13' is true, the ISBN is always
+     * converted to ISBN-13. If both are false (default), then each ISBN is returned
+     * as found in record.
+     *
+     * @return array
+     */
+    public function getCleanISBNs(
```

Comment:
Thanks, @xmorave2! I had a few ideas about additional potentially useful flags and some refactoring to ensure that flags are never unset. Here's my proposed (but currently untested) revision:

```
    public function getCleanISBNs(array $userFlags = []): array
    {
        $defaultFlags = [
             'only10' =>  false,
             'prefer10' => true,
             'filterInvalid' => true,
             'normalize13' => false
        ];
        $flags = $userFlags + $defaultFlags;
        $isbns = $this->getISBNs();
        $all = $tens = $thirteens = $invalid = [];
        foreach ($isbns as $isbn) {
            // Strip off any unwanted notes:
            if ($pos = strpos($isbn, ' ')) {
                $isbn = substr($isbn, 0, $pos);
            }
            $isbnObj = new ISBN($isbn);
            if ($isbnObj->isValid()) {
                if ($isbn10 = $isbnObj->get10()) {
                    $normalized = $flags['normalize13'] ? $isbnObj->get13() : $isbn10;
                    $tens[] = $normalized;
                    $all[] = $normalized;
                } elseif ($isbn13 = $isbnObj->get13()) {
                    $thirteens[] = $isbn13;
                    $all[] = $isbn13;
                } elseif (!$flags['filterInvalid']) {
                    $invalid[] = $isbn;
                    $all[] = $isbn;
                }
            }
        }
        return $flags['prefer10'] ? array_merge($tens, $thirteens, $invalid) : $all;
    }
```

What do you think? This adds the "prefer 10" flag set to true by default so that we can retain backward compatibility with the getCleanISBN logic which favors ISBN10-compatible values in case some APIs are not 13-compatible (probably not a major concern at this point in time, but a bit of compatibility we might as well maintain if we easily can).

Because of all the possible code paths through this, we should add some unit tests to cover this logic. I'm happy to write something next week when I'm back from vacation if that would be helpful.

Reply:
I tried to refactor it this way. 
What do you think about it now?